### PR TITLE
Make scheduled stages order deterministic

### PR DESCRIPTION
**Pull request by sopel39**
The scheduler is single threaded per query so the actual order stages
are processed in the scheduler can affect short queries on fast connectors.

### CODE REVIEW

**Comment by sopel39** [comment]
cc @martint

File: core/trino-main/src/main/java/io/trino/execution/scheduler/policy/PhasedExecutionSchedule.java [hunk 1]
```diff
@@ -79,8 +81,14 @@
      * Graph representing topology between fragments (e.g. child -> parent relationship).
      */
     private final DirectedGraph<PlanFragmentId, FragmentsEdge> fragmentTopology;
+    private final List<PlanFragmentId> sortedFragments = new ArrayList<>();
     private final Map<PlanFragmentId, StageExecution> stagesByFragmentId;
-    private final Set<StageExecution> activeStages = new HashSet<>();
+    private final Set<StageExecution> activeStages = new LinkedHashSet<>();
+
+    /**
```
Comment (losipiuk) [inline core/trino-main/src/main/java/io/trino/execution/scheduler/policy/PhasedExecutionSchedule.java:88]: nit: make it non-javadoc comment

File: core/trino-main/src/main/java/io/trino/execution/scheduler/policy/PhasedExecutionSchedule.java [hunk 2]
```diff
@@ -158,41 +173,50 @@ void schedule()
...
+        return blockedFragments.stream()
                 .flatMap(fragmentId -> fragmentTopology.outgoingEdgesOf(fragmentId).stream())
                 .map(FragmentsEdge::getTarget)
+                .collect(toImmutableSet());
+    }
+
+    private void selectForExecution(Set<PlanFragmentId> fragmentIds)
+    {
+        requireNonNull(fragmentOrdering, "fragmentOrdering is null");
+        fragmentOrdering.sortedCopy(fragmentIds).stream()
```
Comment (losipiuk) [inline core/trino-main/src/main/java/io/trino/execution/scheduler/policy/PhasedExecutionSchedule.java:219]: why explicit copy. You can do `fragmentIds.stream().sorted(fragmentOrdering)` right?

File: core/trino-main/src/main/java/io/trino/execution/scheduler/policy/PhasedExecutionSchedule.java [hunk 3]
```diff
@@ -142,8 +149,16 @@ public boolean isFinished()
...
+        ImmutableSet.Builder<PlanFragmentId> fragmentsToExecute = ImmutableSet.builder();
+        fragmentsToExecute.addAll(removeCompletedStages());
+        fragmentsToExecute.addAll(unblockStagesWithFullOutputBuffer());
+        selectForExecution(fragmentsToExecute.build());
+    }
+
+    @VisibleForTesting
+    List<PlanFragmentId> getSortedFragments()
+    {
+        return sortedFragments;
```
Comment (losipiuk) [inline core/trino-main/src/main/java/io/trino/execution/scheduler/policy/PhasedExecutionSchedule.java:161]: What is the point of asserting on this one. Oredering is arbitrary. And you do assertion on execution order anyway. I would make it private.

**Review by losipiuk — APPROVED** [review]

File: core/trino-main/src/main/java/io/trino/execution/scheduler/policy/PhasedExecutionSchedule.java (hunk 1 above)
Comment (sopel39) [reply core/trino-main/src/main/java/io/trino/execution/scheduler/policy/PhasedExecutionSchedule.java:88]: Then `{@link PhasedExecutionSchedule#init(Collection)}` wouldn't be managed by IntelliJ refactors

File: core/trino-main/src/main/java/io/trino/execution/scheduler/policy/PhasedExecutionSchedule.java (hunk 3 above)
Comment (sopel39) [reply core/trino-main/src/main/java/io/trino/execution/scheduler/policy/PhasedExecutionSchedule.java:161]: I want to make sure that global ordering is as expected. Oredering is not arbitrary as it favors build before probe and upstream stage before downstream.

Comment (losipiuk) [reply core/trino-main/src/main/java/io/trino/execution/scheduler/policy/PhasedExecutionSchedule.java:161]: It is worth a comment in code then maybe. As code which collects ordering sequence is just an in-order tree scan. It is not obvious that abovementioned properties are important.

File: core/trino-main/src/main/java/io/trino/execution/scheduler/policy/PhasedExecutionSchedule.java (hunk 1 above)
Comment (losipiuk) [reply core/trino-main/src/main/java/io/trino/execution/scheduler/policy/PhasedExecutionSchedule.java:88]: Meh - not enough of justification IMO. As a javadoc this comment does not make sense. 
But I will not fight over that.

**Pull request merged**